### ISSUE TITLE
BUG subclass with __numpy_ufunc__ segfaults with recarray comparison

### ISSUE THREAD

**Issue by mhvk** [numpy/numpy#4855]
If I define a `ndarray` subclass that implements `__numpy_ufunc__`, it segfaults whenever it contains a record array and a comparison is done (see below). Note that any direct calls to `ufunc` give the usual `TypeError`, and the segfault can be avoided by defining one's own `__eq__`, `__ne__`.

```
import numpy as np
class MyA(np.ndarray):
    def __numpy_ufunc__(self, ufunc, method, i, inputs, **kwargs):
        return getattr(ufunc, method)(*(input.view(np.ndarray)
                                        for input in inputs), **kwargs)

a = np.arange(12.).reshape(4,3)
ma = a.view(MyA)
ma == ma[0]
# works as expected
ra = a.view(dtype=('f8,f8,f8')).squeeze()
ra == ra[0]
# works as expected: array([ True, False, False, False], dtype=bool)
mra = ra.view(MyA)
mra == ra[0]
# Segmentation fault
```

**Comment by mhvk** [comment]
Checking the 1.10 blockers that turned up by my `__numpy_ufunc__` trials: in ` 1.10.0.dev-256d256`, the code at the top still leads to a segfault.

**Comment by mhvk** [comment]
@ahaldane - would your work on `recarray` also solve these segmentation faults? (FWIW, it still happens with current master, `1.10.0.dev0+83f36c5`

**Comment by ahaldane** [comment]
I don't think so. The two PRs I made that fix segfaults (#5947 and #5636) aren't merged yet, but they have to do with getfield/setfield which aren't used here at first glance. 

Btw, this issue involves "structured arrays" rather than recarrays (recarrays are a special subclass `np.recarray` you have to specifically ask for).

Also, the example code above doesn't segfault for me in 1.9.2 or in current master, in python2 and 3.

**Comment by mhvk** [comment]
@ahaldane - OK, thanks; should have thought of the difference between structured and record arrays.

No surprise it doesn't segfault in 1.9.2, since that didn't have `__numpy_ufunc__`, but are you sure that it works with the current `dev` and with python3? I'm using linux, are you using a different OS?

**Comment by ahaldane** [comment]
Oh, I was mistaken. It does segfault in master, on linux. I must have pasted into the wrong terminal. My PRs don't fix it.

**Comment by charris** [comment]
Fixed by #5964, so closing.

**Comment by mhvk** [comment]
Great that this is solved; thanks, @njsmith, @charris.
